http/src/main/resources/jupyter/gce-init.sh

```diff
@@ -0,0 +1,528 @@
+#!/usr/bin/env bash
+
+# This init script instantiates the tool (e.g. Jupyter) docker images on Google Compute Engine instances created by Leo.
+
+set -e -x
+
+#####################################################################################################
+# Functions
+#####################################################################################################
+
+# Retry a command up to a specific number of times until it exits successfully,
+# with exponential back off. For example:
+#
+#   $ retry 5 echo "Hello"
+#     Hello
+#
+#   $ retry 5 false
+#     Retry 1/5 exited 1, retrying in 2 seconds...
+#     Retry 2/5 exited 1, retrying in 4 seconds...
+#     Retry 3/5 exited 1, retrying in 8 seconds...
+#     Retry 4/5 exited 1, retrying in 16 seconds...
+#     Retry 5/5 exited 1, no more retries left.
+function retry {
+  local retries=$1
+  shift
+
+  for ((i = 1; i <= $retries; i++)); do
+    # run with an 'or' so set -e doesn't abort the bash script on errors
+    exit=0
+    "$@" || exit=$?
+    if [ $exit -eq 0 ]; then
+      return 0
+    fi
+    wait=$((2 ** $i))
+    if [ $i -eq $retries ]; then
+      log "Retry $i/$retries exited $exit, no more retries left."
+      break
+    fi
+    log "Retry $i/$retries exited $exit, retrying in $wait seconds..."
+    sleep $wait
+  done
+  return 1
+}
+
+function log() {
+  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $@"
+}
+
+#####################################################################################################
+# Main starts here. It is composed of three sections:
+#   1. Set up that is NOT specific to GCE_OPERATION
+#   2. Operations we want to perform only when we are 'creating' a VM
+#   3. Operations we want to perform only when we are 'restarting' a VM that was previously created
+#####################################################################################################
+
+# Array for instrumentation
+# UPDATE THIS IF YOU ADD MORE STEPS:
+# currently the steps are:
+# START init,
+# .. after env setup
+# .. after copying files from google and into docker
+# .. after docker compose
+# .. after welder start
+# .. after hail and spark
+# .. after nbextension install
+# .. after server extension install
+# .. after combined extension install
+# .. after user script
+# .. after lab extension install
+# .. after jupyter notebook start
+# END
+
+
+#####################################################################################################
+# Set up that is NOT specific to GCE_OPERATION
+#####################################################################################################
+STEP_TIMINGS=($(date +%s))
```

Comment:
this is an array of milliseconds since epoch to track the time between the points it is invoked. I'd say its worth either removing them, or thinking about where it is relevant to take a snapshot of how long the script has taken so far. I think accompanying each with a log of what 'step' it is currently on, as a lot of them have, would be sufficient if you want to keep them

Reply:
I don't think the current format is very user-friendly. I'll make it report the entire execution time for the time being and if it becomes necessary, we can add more granular execution times later.